drivers/lxc/driver.go

```diff
@@ -0,0 +1,510 @@
+	StartedAt     time.Time
+}
+
+// NewLXCDriver returns a new DriverPlugin implementation
+func NewLXCDriver(logger hclog.Logger) drivers.DriverPlugin {
+	ctx, cancel := context.WithCancel(context.Background())
+	logger = logger.Named(pluginName)
+	return &Driver{
+		eventer:        eventer.NewEventer(ctx, logger),
+		config:         &Config{},
+		tasks:          newTaskStore(),
+		ctx:            ctx,
+		signalShutdown: cancel,
+		logger:         logger,
+	}
+}
+
+func (d *Driver) PluginInfo() (*base.PluginInfoResponse, error) {
+	return pluginInfo, nil
+}
+
+func (d *Driver) ConfigSchema() (*hclspec.Spec, error) {
+	return configSpec, nil
+}
+
+func (d *Driver) SetConfig(data []byte, cfg *base.ClientAgentConfig) error {
+	var config Config
+	if err := base.MsgPackDecode(data, &config); err != nil {
+		return err
+	}
+
+	d.config = &config
+	if cfg != nil {
+		d.nomadConfig = cfg.Driver
+	}
+
+	return nil
+}
+
+func (d *Driver) Shutdown(ctx context.Context) error {
+	d.signalShutdown()
+	return nil
+}
+
+func (d *Driver) TaskConfigSchema() (*hclspec.Spec, error) {
+	return taskConfigSpec, nil
+}
+
+func (d *Driver) Capabilities() (*drivers.Capabilities, error) {
+	return capabilities, nil
+}
+
+func (d *Driver) Fingerprint(ctx context.Context) (<-chan *drivers.Fingerprint, error) {
+	ch := make(chan *drivers.Fingerprint)
+	go d.handleFingerprint(ctx, ch)
+	return ch, nil
+}
+
+func (d *Driver) handleFingerprint(ctx context.Context, ch chan<- *drivers.Fingerprint) {
+	defer close(ch)
+	ticker := time.NewTimer(0)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-d.ctx.Done():
+			return
+		case <-ticker.C:
+			ticker.Reset(fingerprintPeriod)
+			ch <- d.buildFingerprint()
+		}
+	}
+}
+
+func (d *Driver) buildFingerprint() *drivers.Fingerprint {
+	var health drivers.HealthState
+	var desc string
+	attrs := map[string]string{}
+
+	lxcVersion := lxc.Version()
+
+	if d.config.Enabled && lxcVersion != "" {
+		health = drivers.HealthStateHealthy
+		desc = "ready"
+		attrs["driver.lxc"] = "1"
+		attrs["driver.lxc.version"] = lxcVersion
+	} else {
+		health = drivers.HealthStateUndetected
+		desc = "disabled"
+	}
+
+	if d.config.AllowVolumes {
+		attrs["driver.lxc.volumes.enabled"] = "1"
+	}
+
+	return &drivers.Fingerprint{
+		Attributes:        attrs,
+		Health:            health,
+		HealthDescription: desc,
+	}
+}
+
+func (d *Driver) RecoverTask(handle *drivers.TaskHandle) error {
+	if handle == nil {
+		return fmt.Errorf("error: handle cannot be nil")
+	}
+
+	if _, ok := d.tasks.Get(handle.Config.ID); ok {
+		return nil
+	}
+
+	var taskState TaskState
+	if err := handle.GetDriverState(&taskState); err != nil {
+		return fmt.Errorf("failed to decode task state from handle: %v", err)
+	}
+
+	var driverConfig TaskConfig
+	if err := taskState.TaskConfig.DecodeDriverConfig(&driverConfig); err != nil {
+		return fmt.Errorf("failed to decode driver config: %v", err)
+	}
+
+	c, err := lxc.NewContainer(taskState.ContainerName, d.lxcPath())
```

Comment:
Should call this out - I opted to have lxcPath be a property of the driver rather than individual containers.  The implication is that if nomad/driver is restarted with a new lxc path configuration, the driver would lose previous container state.

Not obvious to me why the driver would be restarted with a new lxc path, so not obvious to me what's the least surprising behavior is.  Could be because they want to wipe state and start a new, or could be because they want to transition to new path while keeping old containers intact.

Reply:
I don't think this is the only place we would have inconsistencies if users changed paths between restarts, so it seems fine not to worry about it here.